### PR TITLE
Ship Stargz Snapshotter

### PR DESCRIPTION
**Pull request by ktock**
#### Proposed Changes ####

This commit adds stargz snapshotter to k3s based on the discussion https://github.com/k3s-io/k3s/issues/2918. Stargz Snapshotter can be enabled by `--snapshotter=stargz` option.

#### Types of Changes ####

New Feature

#### Verification ####

`k3s server --snapshotter=stargz`

~~`/dev/fuse` and `fusermount` FUSE helper binary are required on the host.~~

#### Linked Issues ####

https://github.com/k3s-io/k3s/issues/2918

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -68,8 +69,9 @@ require (
 	github.com/NYTimes/gziphandler v1.1.1 // indirect
 	github.com/bronze1man/goStrongswanVici v0.0.0-20190828090544-27d02f80ba40 // indirect
 	github.com/containerd/cgroups v0.0.0-20200710171044-318312a37340
-	github.com/containerd/containerd v1.4.1
+	github.com/containerd/containerd v1.5.0-beta.0.0.20210122062454-5a66c2ae5cec
```
Comment (AkihiroSuda) [inline go.mod:72]: github.com/k3s-io/containerd  needs to be synced for this?

Comment (ktock) [reply go.mod:72]: Stargz Snapshotter v0.4.1 is tested with containerd v1.4.3 so `github.com/k3s-io/containerd v1.4.3-k3s3` seems fine.

https://github.com/containerd/stargz-snapshotter/blob/v0.4.1/Dockerfile#L15

File: go.mod [hunk 2]
```diff
@@ -68,8 +69,9 @@ require (
...
 	github.com/containerd/cri v1.11.1-0.20200820101445-b0cc07999aa5
+	github.com/containerd/stargz-snapshotter v0.4.2-0.20210215092109-c97b82620fb2
```
Comment (AkihiroSuda) [inline go.mod:74]: Can we have a new release soon for this?

Comment (ktock) [reply go.mod:74]: Yes we have.

**Comment by dweomer** [comment]
@ktock (and @AkihiroSuda) thank you for this! Is there an automated way to assert that an image is being pulled lazily?

**Comment by ktock** [comment]
Lazy layers (snapshots) contain the label `"containerd.io/snapshot/remote"="remote snapshot"` like the following:

```console
# ${K3S} ctr --namespace=k8s.io snapshot --snapshotter=stargz info sha256:d6a476f6c79bafc0ace7c32eda6e27c9c022f0b53e67b33fa8acdcc184ddddb2 | jq '.Labels."containerd.io/snapshot/remote"'
"remote snapshot"
```

So we can check whether an image is lazy, by testing snapshots of this image contain that label.

The shellscript doing this will be something like the following:

```bash
#!/bin/bash

set -euo pipefail

REF=${1}

K3S=${K3S:-k3s}
FAIL=false
LAYERSNUM=0
IMAGE=$(${K3S} ctr --namespace=k8s.io image ls name=="${REF}" | grep "${REF}" | cut -f 3 -d ' ')
${K3S} ctr --namespace=k8s.io content get ${IMAGE} | jq -r '.manifests[].digest' | while read MANIFEST ; do
    CONFIG=$(${K3S} ctr --namespace=k8s.io content get ${MANIFEST} | jq -r '.config.digest')
    SNAPSHOT=$(${K3S} ctr --namespace=k8s.io content ls digest==${CONFIG} | grep containerd.io/gc.ref.snapshot.stargz | sed -E 's|.*containerd\.io/gc\.ref\.snapshot\.stargz=([0-9a-z:]*).*|\1|g')
    for (( ; ; )) ; do
        LABEL=$(${K3S} ctr --namespace=k8s.io snapshot --snapshotter=stargz info ${SNAPSHOT} | jq -r '.Labels."containerd.io/snapshot/remote"')
        if [ "${LABEL}" == "remote snapshot" ] ; then
            echo "snapshot ${SNAPSHOT} is lazy"
        else
            echo "snapshot ${SNAPSHOT} is NOT lazy"
            FAIL=true
        fi
        SNAPSHOT=$(${K3S} ctr --namespace=k8s.io snapshot --snapshotter=stargz info ${SNAPSHOT} | jq -r '.Parent')
        if [ "${SNAPSHOT}" == "null" ] ; then
            break
        elif [ ${LAYERSNUM} -gt 100 ] ; then
            echo "testing image contains too many layes > 100"
            exit 1
        fi
        ((LAYERSNUM+=1))
    done
done

if [ "${FAIL}" != "false" ] ; then
    echo "some layers aren't lazy"
    exit 1
fi
```

```console
# /test.sh ghcr.io/stargz-containers/python:3.7-esgz
snapshot sha256:d6a476f6c79bafc0ace7c32eda6e27c9c022f0b53e67b33fa8acdcc184ddddb2 is lazy
snapshot sha256:b2063a3092fe8636f2edbe9f3a48d7f12397bab33f2203523aa0fbb7e1b49da3 is lazy
snapshot sha256:bec907704d956aa3de19fd6e64503cecedb6a880e7e7c912bff08e95e9d043c3 is lazy
snapshot sha256:be70ca3c5b2bc8cdb6661bd0f6e5a9af4b9cac3fa5d92459160a9f080b3f1a43 is lazy
snapshot sha256:ce4fe4e9a134692c19de701ff8bd1a312d69067f2627c7ab3269d8aaa13b45cf is lazy
snapshot sha256:0781e2aaf77fada02417c1535fd8f9bee2712c80b247992f5cdf9fb584745bfc is lazy
snapshot sha256:90ad6c7ecfb79e485dc7166f3ad8966b0a3256ac45891f5bd5df0bf4c70d2f6c is lazy
snapshot sha256:a6edcd7520ff8464afbecff28848096cdb28f15cebbdb72117c137a1c4d84b2a is lazy
snapshot sha256:c009c03d02ca3e7b6f9a437d6b0b3fef7191a3794703d7a3297b0c879c2ab8d4 is lazy
```

**Comment by dweomer** [comment]
Some notes from an internal meeting where this was discussed:
- before we merge this we would like a use-case demo (a good candidate for upcoming hack-week + blog)
- we would also want a docs pr that talks about the stargz snapshotter and the why/how of using it (config + preprocessing concerns. the ntt article has some material that we could mine)
- automated tests/assertions for correctness (we have a start on this [in an earlier comment](https://github.com/k3s-io/k3s/pull/2936#issuecomment-780248523))

These items imply work and I want to be explicit: @ktock we appreciate the contribution (and quick turn-around!) and don't expect you to do this work. That said, contributions are always welcome!

/cc @cjellick

**Comment by xiaods** [comment]
> Some notes from an internal meeting where this was discussed:
> 
> * before we merge this we would like a use-case demo (a good candidate for upcoming hack-week + blog)
> * we would also want a docs pr that talks about the stargz snapshotter and the why/how of using it (config + preprocessing concerns. the ntt article has some material that we could mine)
> * automated tests/assertions for correctness (we have a start on this [in an earlier comment](https://github.com/k3s-io/k3s/pull/2936#issuecomment-780248523))
> 
> These items imply work and I want to be explicit: @ktock we appreciate the contribution (and quick turn-around!) and don't expect you to do this work. That said, contributions are always welcome!
> 
> /cc @cjellick

please hold some time and waiting some new demo insight.

**Comment by ktock** [comment]
@dweomer Thank you for your comments. I've worked on improvements on stargz snapshotter's side especially for the integration of registry configuration. And recently I'm restarting to work on this patch.

> - we would also want a docs pr that talks about the stargz snapshotter and the why/how of using it (config + preprocessing concerns. the ntt article has some material that we could mine)

Opened PR https://github.com/rancher/docs/pull/3322.

> - automated tests/assertions for correctness (we have a start on this in an earlier comment)

Added this test.

> - before we merge this we would like a use-case demo (a good candidate for upcoming hack-week + blog)

As a demo, I'm trying to integrate lazy pulling with Argo Workflow. For the workflow that uses large images, lazy pulling can hopefully shorten the time to complete.

I've created [a small example of Golang CI on k3s (k3d) + Argo Workflow](https://github.com/ktock/stargz-snapshotter/blob/k3sdemo/script/k3sdemo/go.yaml.template). On the cold start case, lazy pulling shortens the time to take for this workflow from 91 sec to 51 sec on average (measured at https://github.com/ktock/stargz-snapshotter/actions/runs/917330760).

Workflow with non-lazy standard images:

```
STEP                                                                  TEMPLATE               PODNAME                      DURATION  MESSAGE
 ✔ ci-example-94swm                                                   ci-example                                                      
 ├───✔ golangci-lint                                                  golangci-lint-example  ci-example-94swm-3369489379  33s         
 ├───✔ gosec                                                          gosec-example          ci-example-94swm-2377571633  11s         
 ├───✔ build                                                          build-golang-example   ci-example-94swm-267001449   17s         
 └─┬─✔ test(0:image:ghcr.io/stargz-containers/alpine,tag:3.10.2-org)  run-hello              ci-example-94swm-2420311356  6s          
   ├─✔ test(1:image:ghcr.io/ktock/ubuntu,tag:20.04-org)               run-hello              ci-example-94swm-2723959033  7s          
   └─✔ test(2:image:ghcr.io/stargz-containers/fedora,tag:30-org)      run-hello              ci-example-94swm-909720077   8s
```

Workflow with eStargz images:

```
STEP                                                                   TEMPLATE               PODNAME                      DURATION  MESSAGE
 ✔ ci-example-2fckt                                                    ci-example                                                      
 ├───✔ golangci-lint                                                   golangci-lint-example  ci-example-2fckt-169023647   18s         
 ├───✔ gosec                                                           gosec-example          ci-example-2fckt-2926929885  6s          
 ├───✔ build                                                           build-golang-example   ci-example-2fckt-4249784069  6s          
 └─┬─✔ test(0:image:ghcr.io/stargz-containers/alpine,tag:3.10.2-esgz)  run-hello              ci-example-2fckt-2351514715  8s          
   ├─✔ test(1:image:ghcr.io/ktock/ubuntu,tag:20.04-esgz)               run-hello              ci-example-2fckt-18456000    8s          
   └─✔ test(2:image:ghcr.io/stargz-containers/fedora,tag:30-esgz)      run-hello              ci-example-2fckt-2112439900  8s
```

I'm very welcome if anyone has better example of workflow where lazy pulling can improve the performance.

**Comment by AkihiroSuda** [comment]
Needs rebase

**Comment by ktock** [comment]
Thanks for notifying. rebased.

**Comment by AkihiroSuda** [comment]
Needs rebase again

**Comment by ktock** [comment]
rebased.

**Comment by AkihiroSuda** [comment]
Needs rebase again

**Comment by dweomer** [comment]
> @cwayne18 @brandond @dweomer - @AkihiroSuda and @ktock are keeping this PR alive. What can we do to move it along?

I want to honor this work by including it as soon as is reasonable for our schedule. So, 1.22.x (one assumes after the current 1.22.1 release cycle) or 1.23.x?

**Comment by ktock** [comment]
Bumped up stargz snapshotter to v0.8.0 and all green.

**Review by dweomer — APPROVED** [review]

**Comment by brandond** [comment]
:ship:  :it:  @dweomer

**Pull request merged**